atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraTimestampUtils.java

```diff
@@ -0,0 +1,145 @@
+    private static final String VALUE_COLUMN = "value";
+
+    private static final byte[] SUCCESSFUL_OPERATION = {1};
+
+    private static final String BACKUP_COLUMN_NAME = "oldTs";
+    private static final byte[] INVALIDATED_VALUE = new byte[1];
+
+    private CassandraTimestampUtils() {
+        // utility class
+    }
+
+    public static ByteBuffer constructSelectTimestampBoundQuery() {
+        String selectQuery = String.format(
+                "SELECT %s FROM %s WHERE key=%s AND column1=%s AND column2=%s;",
+                VALUE_COLUMN,
+                wrapInQuotes(AtlasDbConstants.TIMESTAMP_TABLE.getQualifiedName()),
+                ROW_AND_COLUMN_NAME_HEX_STRING,
+                ROW_AND_COLUMN_NAME_HEX_STRING,
+                CASSANDRA_TIMESTAMP);
+        return ByteBuffer.wrap(PtBytes.toBytes(selectQuery));
+    }
+
+    public static ByteBuffer constructCheckAndSetCqlQuery(Long expected, long target) {
+        String updateQuery = (expected == null)
+                ? constructInsertIfNotExistsQuery(target)
+                : constructUpdateIfEqualQuery(expected, target);
+
+        return ByteBuffer.wrap(PtBytes.toBytes(updateQuery));
+    }
+
+    public static boolean wasOperationApplied(CqlResult cqlResult) {
+        Column appliedColumn = getNamedColumn(cqlResult, APPLIED_COLUMN);
+        return Arrays.equals(appliedColumn.getValue(), SUCCESSFUL_OPERATION);
+    }
+
+    public static long getLongValue(CqlResult cqlResult) {
+        Column valueColumn = getNamedColumn(cqlResult, VALUE_COLUMN);
+        if (Arrays.equals(INVALIDATED_VALUE, valueColumn.getValue())) {
+            throw new IllegalStateException("Couldn't extract a long from the invalidated value!");
+        }
+        return PtBytes.toLong(valueColumn.getValue());
+    }
+
+    private static Column getNamedColumn(CqlResult cqlResult, String columnName) {
+        List<Column> columns = getColumnsFromOnlyRow(cqlResult);
+        return Iterables.getOnlyElement(
+                columns.stream()
+                        .filter(column -> columnName.equals(PtBytes.toString(column.getName())))
+                        .collect(Collectors.toList()));
+    }
+
+    private static String constructUpdateIfEqualQuery(Long expected, long target) {
+        return String.format(
+                "UPDATE %s SET value=%s WHERE key=%s AND column1=%s AND column2=%s IF value=%s;",
+                wrapInQuotes(AtlasDbConstants.TIMESTAMP_TABLE.getQualifiedName()),
+                encodeCassandraHexBytes(PtBytes.toBytes(target)),
+                ROW_AND_COLUMN_NAME_HEX_STRING,
+                ROW_AND_COLUMN_NAME_HEX_STRING,
+                CASSANDRA_TIMESTAMP,
+                encodeCassandraHexBytes(PtBytes.toBytes(expected)));
+    }
+
+    private static String constructInsertIfNotExistsQuery(long target) {
+        return String.format(
+                "INSERT INTO %s (key, column1, column2, value) VALUES (%s, %s, %s, %s) IF NOT EXISTS;",
```

Comment:
The Cassandra client pool will return a result that has an "[applied]" column. So if the value already exists, we'll get a `CqlResponse` which has a mapping something like

```java
 [applied] | key    | column1 | column2 | value
-----------+--------+---------+---------+--------------------
     False | 0x7473 |  0x7473 |      -1 | 0x00000000000f6951
```

(for CAS, it returns the value in the DB; for PUE it returns the existing record).

When we store a timestamp bound, the CassandraTimestampStore will check the [applied] column, and throw a `ConcurrentModificationException` if the CAS failed. CTBS will catch the exception, wrap it in a MRTSE and throw that.